File: frontend/src/services/productService.js
import { productApi } from './api'

// Get all products
export const getAllProducts = () =>
  productApi.get('/api/products')

// Get a single product by ID
export const getProductById = (id) =>
  productApi.get(`/api/products/${id}`)

// Search products by name (partial match)
export const searchProducts = (name) =>
  productApi.get(`/api/products/search?name=${encodeURIComponent(name)}`)

// Get all products in a category
export const getProductsByCategory = (category) =>
  productApi.get(`/api/products/category/${encodeURIComponent(category)}`)

// Create a new product (admin)
export const createProduct = (data) =>
  productApi.post('/api/products', data)

// Update a product (admin)
export const updateProduct = (id, data) =>
  productApi.put(`/api/products/${id}`, data)

// Delete a product (admin)
export const deleteProduct = (id) =>
  productApi.delete(`/api/products/${id}`)

File: backend/product-service/src/main/java/com/ecommerce/productservice/service/ProductService.java
package com.ecommerce.productservice.service;

import com.ecommerce.productservice.dto.CreateProductRequest;
import com.ecommerce.productservice.dto.ProductDTO;

import java.util.List;

/**
 * ProductService interface - defines all product-related business operations.
 */
public interface ProductService {

    /** Add a new product to the catalog */
    ProductDTO createProduct(CreateProductRequest request);

    /** Get a single product by ID */
    ProductDTO getProductById(Long id);

    /** Get all products */
    List<ProductDTO> getAllProducts();

    /** Search products by name (partial match) */
    List<ProductDTO> searchByName(String name);

    /** Get products filtered by category */
    List<ProductDTO> getByCategory(String category);

    /** Update an existing product */
    ProductDTO updateProduct(Long id, CreateProductRequest request);

    /** Remove a product from the catalog */
    void deleteProduct(Long id);
}
